src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTimingTest.java

```diff
@@ -876,79 +813,139 @@ public void parallel() throws Exception {
+        assertEquals(2, visitor.chunks.size());
+        visitor.assertStageOrBranchStatus("succeeds", GenericStatus.SUCCESS);
+        visitor.assertStageOrBranchStatus("fails", GenericStatus.FAILURE);
     }
 
-    private static List<MemoryFlowChunk> getStages(FlowExecution execution) {
-        StageTest.CollectingChunkVisitor visitor = new StageTest.CollectingChunkVisitor();
-        ForkScanner.visitSimpleChunks(execution.getCurrentHeads(), visitor, new StageChunkFinder());
-        return visitor.getChunks();
+    @Test
+    @Issue("JENKINS-39203")
+    public void unstableWithWarningAction() throws Exception {
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "unstable");
+        job.setDefinition(new CpsFlowDefinition(
+                "stage('unstable') {\n" +
+                "  echo('foo')\n" +
+                "  unstable('oops')\n" +
+                "  echo('foo')\n" +
+                "}\n" +
+                "stage('success') {\n" +
+                "  echo('no problem')" +
+                "}\n" +
+                "stage('failure') {\n" +
+                "  unstable('second oops')\n" +
+                "  error('failure')\n" +
+                "}\n", true));
+        WorkflowRun run = j.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0));
+        StagesAndParallelBranchesVisitor visitor = new StagesAndParallelBranchesVisitor(run);
+        assertEquals(3, visitor.chunks.size());
+        visitor.assertStageOrBranchStatus("unstable", GenericStatus.UNSTABLE);
+        visitor.assertStageOrBranchStatus("success", GenericStatus.SUCCESS);
+        visitor.assertStageOrBranchStatus("failure", GenericStatus.FAILURE);
     }
 
-    private static List<MemoryFlowChunk> getParallelBranches(FlowExecution execution) {
-        CollectingParallelBranchesVisitor visitor = new CollectingParallelBranchesVisitor();
-        ForkScanner.visitSimpleChunks(execution.getCurrentHeads(), visitor, new StageChunkFinder());
-        return visitor.getParallelBranches();
+    @Test
+    @Issue("JENKINS-39203")
+    public void unstableInBlockScopeStep() throws Exception {
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "unstable");
+        job.setDefinition(new CpsFlowDefinition(
+                "stage('unstable') {\n" +
+                "  timeout(1) {\n" +
+                "    timeout(1) {\n" +
+                "      unstable('oops')\n" +
+                "    }\n" +
+                "  }\n" +
+                "}\n", true));
+        WorkflowRun run = j.assertBuildStatus(Result.UNSTABLE, job.scheduleBuild2(0));
+        StagesAndParallelBranchesVisitor visitor = new StagesAndParallelBranchesVisitor(run);
+        assertEquals(1, visitor.chunks.size());
+        visitor.assertStageOrBranchStatus("unstable", GenericStatus.UNSTABLE);
+    }
+
+    @Test
+    @Ignore
+    public void nestedStageParentStatus() throws Exception {
```

Comment:
Demonstrates that `StandardChunkVisitor` doesn't work for nested stages. Doesn't matter in practice because we don't show the status of parent stages in Blue Ocean, but seems worth documenting since Blue Ocean's visitor does not address the bug either.